### ISSUE TITLE
fractional scaling: rounding error

### ISSUE THREAD

**Issue by MaxVerevkin** [riverwm/river#953]
Steps to reproduce:
- Set scaling to 1.25.
- Create a small floating view which supports fractional scaling (I will use `alacritty`, from my testing it correctly follows the protocol).
- Move it around in steps of one pixel (`riverctl move left 1`).
- Observe that sometimes the contents of the view are blurred and sometimes not. This looks like a texture is stretched or compressed by one pixel.

**Comment by ifreund** [comment]
This is the result of the wlroots scene graph using position dependent rounding when scaling: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/a0b1329ee6cce22aca0c9b064fe9c136cc05d630/types/scene/wlr_scene.c#L887

I don't recall if there are reasonable alternative approaches here or not.

**Comment by ifreund** [comment]
Upstream issue https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3793

**Comment by MaxVerevkin** [comment]
Yeah, I tried this patch locally and don't see any blurriness with 1.25 scale:

```diff
diff --git a/types/scene/wlr_scene.c b/types/scene/wlr_scene.c
index f81a8613..58c14199 100644
--- a/types/scene/wlr_scene.c
+++ b/types/scene/wlr_scene.c
@@ -891,7 +891,7 @@ static void scene_node_get_size(struct wlr_scene_node *node,
 }
 
 static int scale_length(int length, int offset, float scale) {
-	return round((offset + length) * scale) - round(offset * scale);
+	return round(length * scale);
 }
 
 static void scale_box(struct wlr_box *box, float scale) {
```

Combined with https://github.com/riverwm/river/issues/816#issuecomment-1877677790 I get crisp image with 1.2 too.

**Comment by ifreund** [comment]
> Yeah, I tried this patch locally and don't see any blurriness with 1.25 scale:

The problem is that your patch trades blurriness for gaps between nodes in the scene in certain positions. I suspect you should be able to see that pretty easily by configuring highly contrasting terminal, border, and background colors and moving a terminal around one pixel at a time.

**Comment by MaxVerevkin** [comment]
> The problem is that your patch trades blurriness for gaps between nodes in the scene in certain positions

Yes, the border is not perfect. But this is a trade I'm willing to make wile there is no better solution :)

**Comment by MaxVerevkin** [comment]
Another way to "fix" it is to not use linear filtering when texture size is almost equal to the target size. This will sometimes result in a row or column of duplicated pixels but IMO it's much better that what we have now. I'm not sure if scene API has an option for this.

**Comment by ifreund** [comment]
> Another way to "fix" it is to not use linear filtering when texture size is almost equal to the target size. This will sometimes result in a row or column of duplicated pixels but IMO it's much better than what we have now. I'm not sure if scene API has an option for this.

Yeah, that's one of the options that was brought up in the #wlroots IRC channel yesterday. wlr_scene does have per-buffer configurable filtering with bilinear and nearest neighbor options.

I haven't looked into automatically setting scene buffers to use nearest neighbor if fractional-scale-v1 is in use though.

**Comment by alex-courtis** [comment]
This is problematic. It surfaced explicitly when sway added xdg fractional scaling support.

After much experimentation with different wlroots versions and compositors there appeared no consistency in rounding vs truncating.

In the end, we came up with scaling to a multiple of 1/8 as the only reliable mechanism. See comments RE 1/120 xdg vs 1/256 wlroots: https://github.com/alex-courtis/way-displays/pull/145/files#diff-8c1507342bbb429d560dfcfefcbce19993e16b9971fc56e8e49849934b0aca5a

**Comment by ifreund** [comment]
It seems to me that a new alternative scale request should be added to `wlr-output-management-v1` that uses a 1/120 fraction to set the scale matching the `wp-fractional-scale-v1` protocol.

**Comment by alex-courtis** [comment]
> It seems to me that a new alternative scale request should be added to `wlr-output-management-v1` that uses a 1/120 fraction to set the scale matching the `wp-fractional-scale-v1` protocol.

Only fly in the ointment is that the existing wlr scale method takes a 256ths `wl_fixed`. We'd need to make it an integer numerator over 120 as per the xdg scale.

**Comment by ifreund** [comment]
> Only fly in the ointment is that the existing wlr scale method takes a 256ths `wl_fixed`. We'd need to make it an integer numerator over 120 as per the xdg scale.

I'm not sure I understand what you're saying. I proposed deprecating the current scale event/request which uses wl_fixed and adding new ones using the same approach as fractional-scale-v1.

Aside, "xdg scale" is meaningless, the fractional scale protocol has the `wp` prefix and is called either `wp-fractional-scale-v1` or just `fractional-scale-v1`.
